File: plugin/src/test/java/com/oldstats/tracking/farming/FarmingDataTest.java
package com.oldstats.tracking.farming;

import static com.oldstats.tracking.farming.PatchDecode.decodePatchState;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertNull;
import static org.junit.Assert.assertTrue;

import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;
import org.junit.Test;

public class FarmingDataTest {

    @Test
    public void everyRegionHasAtLeastOnePatch() {
        for (RegionDef region : FarmingWorldData.regions) {
            assertTrue(region.name + " has no patches", !region.patches.isEmpty());
        }
    }

    @Test
    public void catherbyRegionResolvesAndHasTheExpectedPatchSet() {
        List<RegionDef> candidates = FarmingWorldData.regionsById.get(11062);
        assertEquals(1, candidates.size());
        RegionDef region = candidates.get(0);
        assertEquals("Catherby", region.name);
        Set<PatchImplementation> implementations = region.patches.stream()
            .map(p -> p.implementation)
            .collect(Collectors.toSet());
        assertEquals(
            Set.of(PatchImplementation.ALLOTMENT, PatchImplementation.FLOWER, PatchImplementation.HERB, PatchImplementation.COMPOST),
            implementations
        );
    }

    @Test
    public void catherbyNorthAndSouthAllotmentDisambiguationUsesBoundsCheck() {
        List<RegionDef> candidates = FarmingWorldData.regionsById.get(11317);
        assertEquals(2, candidates.size());

        // Inside the allotment/herb/flower area (south of y=3440 within the x>=2816 band)
        RegionDef allotmentRegion = candidates.stream()
            .filter(r -> r.boundsCheck != null && r.boundsCheck.test(2820, 3441, 0))
            .findFirst()
            .orElseThrow();
        assertTrue(allotmentRegion.patches.stream().anyMatch(p -> p.implementation == PatchImplementation.HERB));

        // The fruit tree patch sits north of that line
        RegionDef fruitTreeRegion = candidates.stream()
            .filter(r -> r.boundsCheck != null && r.boundsCheck.test(2820, 3300, 0))
            .findFirst()
            .orElseThrow();
        assertTrue(fruitTreeRegion.patches.stream().anyMatch(p -> p.implementation == PatchImplementation.FRUIT_TREE));
    }

    @Test
    public void faladorAllotmentOnlyResolvesSouthOfTheYBoundsCheck() {
        List<RegionDef> candidates = FarmingWorldData.regionsById.get(12083);
        RegionDef region = candidates.stream()
            .filter(r -> r.patches.stream().anyMatch(p -> p.implementation == PatchImplementation.ALLOTMENT))
            .findFirst()
            .orElseThrow();
        assertEquals(true, region.boundsCheck.test(0, 3272, 0));
        assertEquals(false, region.boundsCheck.test(0, 3271, 0));
    }

    @Test
    public void farmingGuildHasThirteenDistinctPatches() {
        List<RegionDef> candidates = FarmingWorldData.regionsById.get(4922);
        assertEquals(1, candidates.size());
        assertEquals(13, candidates.get(0).patches.size());
    }

    @Test
    public void herbPatchDecodesRanarrGrowingAndHarvestableCorrectly() {
        // Ranarr: growing 32-35, harvestable 36-38 (see PatchDecode.java / real HERB table)
        assertEquals(new PatchState(Produce.RANARR, CropState.GROWING), decodePatchState(PatchImplementation.HERB, 33));
        assertEquals(new PatchState(Produce.RANARR, CropState.HARVESTABLE), decodePatchState(PatchImplementation.HERB, 37));
        assertEquals(new PatchState(Produce.RANARR, CropState.DISEASED), decodePatchState(PatchImplementation.HERB, 141));
    }

    @Test
    public void allotmentPatchDecodesPotatoLifecycleCorrectly() {
        assertEquals(new PatchState(Produce.WEEDS, CropState.GROWING), decodePatchState(PatchImplementation.ALLOTMENT, 0));
        assertEquals(new PatchState(Produce.POTATO, CropState.GROWING), decodePatchState(PatchImplementation.ALLOTMENT, 7));
        assertEquals(new PatchState(Produce.POTATO, CropState.HARVESTABLE), decodePatchState(PatchImplementation.ALLOTMENT, 11));
        assertEquals(new PatchState(Produce.POTATO, CropState.DISEASED), decodePatchState(PatchImplementation.ALLOTMENT, 136));
        assertEquals(new PatchState(Produce.POTATO, CropState.DEAD), decodePatchState(PatchImplementation.ALLOTMENT, 200));
    }

    @Test
    public void compostBinDecodesEmptyAndFillingCorrectly() {
        assertEquals(new PatchState(Produce.EMPTY_COMPOST_BIN, CropState.EMPTY), decodePatchState(PatchImplementation.COMPOST, 0));
        assertEquals(new PatchState(Produce.COMPOST, CropState.FILLING), decodePatchState(PatchImplementation.COMPOST, 5));
        assertEquals(new PatchState(Produce.COMPOST, CropState.HARVESTABLE), decodePatchState(PatchImplementation.COMPOST, 20));
    }

    @Test
    public void everyPatchVarbitReferencedByARegionHasDecodeCoverageForAtLeastValue0() {
        for (RegionDef region : FarmingWorldData.regions) {
            for (PatchDef patch : region.patches) {
                assertNotNull(
                    region.name + " " + patch.name + " (" + patch.implementation + ") has no decode for value 0",
                    decodePatchState(patch.implementation, 0)
                );
            }
        }
    }

    @Test
    public void unknownRegionResolvesToNothing() {
        assertNull(FarmingWorldData.regionsById.get(999999));
    }
}

File: plugin/src/test/kotlin/com/oldstats/tracking/farming/FarmingDataTest.kt
package com.oldstats.tracking.farming

import org.junit.Assert.assertEquals
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test

class FarmingDataTest {

    @Test
    fun `every region has at least one patch`() {
        for (region in FarmingWorldData.regions) {
            assertTrue("${region.name} has no patches", region.patches.isNotEmpty())
        }
    }

    @Test
    fun `catherby region resolves and has the expected patch set`() {
        val candidates = FarmingWorldData.regionsById.getValue(11062)
        val region = candidates.single()
        assertEquals("Catherby", region.name)
        val implementations = region.patches.map { it.implementation }.toSet()
        assertEquals(
            setOf(PatchImplementation.ALLOTMENT, PatchImplementation.FLOWER, PatchImplementation.HERB, PatchImplementation.COMPOST),
            implementations,
        )
    }

    @Test
    fun `catherby north and south allotment disambiguation uses bounds check`() {
        val candidates = FarmingWorldData.regionsById.getValue(11317)
        assertEquals(2, candidates.size)

        // Inside the allotment/herb/flower area (south of y=3440 within the x>=2816 band)
        val allotmentRegion = candidates.first { it.boundsCheck?.invoke(2820, 3441, 0) == true }
        assertTrue(allotmentRegion.patches.any { it.implementation == PatchImplementation.HERB })

        // The fruit tree patch sits north of that line
        val fruitTreeRegion = candidates.first { it.boundsCheck?.invoke(2820, 3300, 0) == true }
        assertTrue(fruitTreeRegion.patches.any { it.implementation == PatchImplementation.FRUIT_TREE })
    }

    @Test
    fun `falador allotment only resolves south of the y bounds check`() {
        val candidates = FarmingWorldData.regionsById.getValue(12083)
        val region = candidates.single { it.patches.any { p -> p.implementation == PatchImplementation.ALLOTMENT } }
        assertEquals(true, region.boundsCheck?.invoke(0, 3272, 0))
        assertEquals(false, region.boundsCheck?.invoke(0, 3271, 0))
    }

    @Test
    fun `farming guild has thirteen distinct patches`() {
        val region = FarmingWorldData.regionsById.getValue(4922).single()
        assertEquals(13, region.patches.size)
    }

    @Test
    fun `herb patch decodes ranarr growing and harvestable correctly`() {
        // Ranarr: growing 32-35, harvestable 36-38 (see PatchDecode.kt / real HERB table)
        assertEquals(PatchState(Produce.RANARR, CropState.GROWING), decodePatchState(PatchImplementation.HERB, 33))
        assertEquals(PatchState(Produce.RANARR, CropState.HARVESTABLE), decodePatchState(PatchImplementation.HERB, 37))
        assertEquals(PatchState(Produce.RANARR, CropState.DISEASED), decodePatchState(PatchImplementation.HERB, 141))
    }

    @Test
    fun `allotment patch decodes potato lifecycle correctly`() {
        assertEquals(PatchState(Produce.WEEDS, CropState.GROWING), decodePatchState(PatchImplementation.ALLOTMENT, 0))
        assertEquals(PatchState(Produce.POTATO, CropState.GROWING), decodePatchState(PatchImplementation.ALLOTMENT, 7))
        assertEquals(PatchState(Produce.POTATO, CropState.HARVESTABLE), decodePatchState(PatchImplementation.ALLOTMENT, 11))
        assertEquals(PatchState(Produce.POTATO, CropState.DISEASED), decodePatchState(PatchImplementation.ALLOTMENT, 136))
        assertEquals(PatchState(Produce.POTATO, CropState.DEAD), decodePatchState(PatchImplementation.ALLOTMENT, 200))
    }

    @Test
    fun `compost bin decodes empty and filling correctly`() {
        assertEquals(PatchState(Produce.EMPTY_COMPOST_BIN, CropState.EMPTY), decodePatchState(PatchImplementation.COMPOST, 0))
        assertEquals(PatchState(Produce.COMPOST, CropState.FILLING), decodePatchState(PatchImplementation.COMPOST, 5))
        assertEquals(PatchState(Produce.COMPOST, CropState.HARVESTABLE), decodePatchState(PatchImplementation.COMPOST, 20))
    }

    @Test
    fun `every patch varbit referenced by a region has decode coverage for at least value 0`() {
        for (region in FarmingWorldData.regions) {
            for (patch in region.patches) {
                assertNotNull(
                    "${region.name} ${patch.name} (${patch.implementation}) has no decode for value 0",
                    decodePatchState(patch.implementation, 0),
                )
            }
        }
    }

    @Test
    fun `unknown region resolves to nothing`() {
        assertNull(FarmingWorldData.regionsById[999999])
    }
}
